File: src/bin/lg-uva437.cc
#include<bits/stdc++.h>
using namespace std;
using ll = int64_t;

ll solve(vector<vector<ll>> boxes) {
    int m = boxes.size();
    int n = m * 3;
    for (int i = 0; i < m; ++i) {
        boxes.push_back({boxes[i][0], boxes[i][2], boxes[i][1]});
        boxes.push_back({boxes[i][1], boxes[i][2], boxes[i][0]});
    }
    vector<vector<int>> ch(n);
    vector<int> ind(n, 0);
    function<bool(const vector<ll>&, const vector<ll>&)> is_fit = [](const vector<ll>& x, const vector<ll>& y) {
        return x[0] < y[0] && x[1] < y[1] || x[0] < y[1] && x[1] < y[0];
    };
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n ;++j) {
            if (is_fit(boxes[j], boxes[i])) {
                ch[i].push_back(j);
                ++ind[j];
            }
        }
    }
    ll res = 0;
    deque<int> dq;
    vector<ll> dp(n, 0);
    for (int i = 0; i < n; ++i) {
        if (!ind[i]) dq.push_back(i);
    }
    while (!dq.empty()) {
        int i = dq.front();
        dq.pop_front();
        dp[i] += boxes[i][2];
        res = max(res, dp[i]);
        for (auto&& j : ch[i]) {
            dp[j] = max(dp[j], dp[i]);
            --ind[j];
            if (!ind[j]) dq.push_back(j);
        }
    }
    return res;
}

int main() {
    int i = 0;
    while (1) {
        ++i;
        int n;
        cin >> n;
        if (!n) break;
        vector<vector<ll>> boxes(n);
        for (int i = 0;i < n; ++i) {
            for (int j = 0; j < 3; ++j) {
                int t; cin >> t; boxes[i].push_back(t);
            }
        }
        cout << "Case " << i << ": maximum height = " << solve(move(boxes)) << endl;
    }
}

File: src/bin/lg-uva437.rs
use std::collections::VecDeque;

fn solve(mut boxes: Vec<Vec<i64>>) -> i64 {
    let m = boxes.len();
    let n = m * 3;
    for i in 0..m {
        let curr = boxes[i].to_owned();
        boxes.push(vec![curr[0], curr[2], curr[1]]);
        boxes.push(vec![curr[1], curr[2], curr[0]]);
    }
    let mut ch = vec![vec![];n];
    let mut ind = vec![0;n];
    let is_fit = |x: &Vec<i64>, y: &Vec<i64>| {
        x[0] < y[0] && x[1] < y[1] || x[0] < y[1] && x[1] < y[0]
    };
    for i in 0..n {
        for j in 0..n {
            if is_fit(&boxes[j], &boxes[i]) {
                ch[i].push(j);
                ind[j] += 1;
            }
        }
    }
    let mut res = 0;
    let mut dq = VecDeque::new();
    let mut dp = vec![0;n];
    for i in 0..n {
        if ind[i] == 0 {
            dq.push_back(i);
        }
    }
    while let Some(i) = dq.pop_front() {
        dp[i] += boxes[i][2];
        res = res.max(dp[i]);
        for &j in &ch[i] {
            dp[j] = dp[j].max(dp[i]);
            ind[j] -= 1;
            if ind[j] == 0 {
                dq.push_back(j);
            }
        }
    }
    res
}

fn main() {
    let mut i = 0;
    loop {
        i += 1;
        let mut buf = String::new();
        std::io::stdin().read_line(&mut buf).ok();
        let n: usize = buf.trim().parse::<usize>().unwrap();
        if n == 0 {
            break;
        }
        let mut boxes = Vec::with_capacity(n);
        for _ in 0..n {
            buf.clear();
            std::io::stdin().read_line(&mut buf).ok();
            let v: Vec<i64> = buf.trim().split(' ').map(|x|x.parse::<i64>().unwrap()).collect::<Vec<i64>>();
            boxes.push(v);
        }
        println!("Case {i}: maximum height = {}", solve(boxes));
    }
}